x/gamm/types/pool.go

```diff
@@ -59,9 +59,13 @@ type PoolI interface {
 	// Pools are expected to guarantee LP'ing at the exact ratio, and single sided LP'ing.
 	JoinPool(ctx sdk.Context, tokensIn sdk.Coins, swapFee sdk.Dec) (numShares sdk.Int, err error)
 
-	// CalcJoinPoolShares returns how many LP shares JoinPool would return on these arguments.
+	// CalcSingleTokenInJoinPoolShares returns how many LP shares JoinPool would return on these arguments( single token join).
 	// This does not mutate the pool, or state.
-	CalcJoinPoolShares(ctx sdk.Context, tokensIn sdk.Coins, swapFee sdk.Dec) (numShares sdk.Int, newLiquidity sdk.Coins, err error)
+	CalcSingleTokenInJoinPoolShares(ctx sdk.Context, tokensIn sdk.Coins, swapFee sdk.Dec) (numShares sdk.Int, newLiquidity sdk.Coins, err error)
+
+	// CalcMultiTokenInJoinPoolShares returns how many LP shares JoinPool would return on these arguments( multi tokens join).
+	// This does not mutate the pool, or state.
+	CalcMultiTokenInJoinPoolShares(ctx sdk.Context, tokensIn sdk.Coins, swapFee sdk.Dec) (numShares sdk.Int, newLiquidity sdk.Coins, err error)
```

Comment:
hrmm, what if we kept the same `CalcJoinPoolShares` API, but then made the first lines of the function doing the split you wrote

Reply:
https://github.com/osmosis-labs/osmosis/pull/1854/commits/909568221759a11b14cc69ca25f8fcc4039134bd
should revert to this commit?

Reply:
I agree with @ValarDragon on adding two functions for `CalcJoinPoolShares` to split into based on its input, and I think splitting into single-token and multi-token joins would be be appropriate abstractions

I'm personally not convinced on the purpose of removing and replacing `CalcJoinPoolShares` altogether though, as it seems to me like we would just be floating the same logic in it up to `JoinPool`, which is already complex enough as it is

So +1 for keeping `CalcJoinPoolShares` and making it internally route to `CalcSingleTokenInJoinPoolShares` and `CalcMultiTokenInJoinPoolShares` based on the passed in token set

Reply:
https://github.com/osmosis-labs/osmosis/pull/1854/commits/1be8c4b359c2e395cde657403711424ea3e89629
Can u check this commit

Reply:
Thanks for the suggestion